services/server/src/server/services/compiler/local/vyperCompiler.ts

```diff
@@ -0,0 +1,214 @@
+  if (process.platform === "darwin") {
+    return "darwin";
+  }
+  if (process.platform === "linux" && process.arch === "x64") {
+    return "linux";
+  }
+  if (process.platform === "win32" && process.arch === "x64") {
+    return "windows.exe";
+  }
+  return false;
+}
+/**
+ * Searches for a vyper compiler: first for a local executable version, then from HOST_VYPER_REPO
+ * Once the compiler is retrieved, it is used, and the stringified vyper output is returned.
+ *
+ * @param version the version of vyper to be used for compilation
+ * @param input a JSON object of the standard-json format compatible with vyper
+ * @param log the logger
+ * @returns stringified vyper output
+ */
+
+export async function useVyperCompiler(
+  vyperRepoPath: string,
+  version: string,
+  vyperJsonInput: VyperJsonInput,
+): Promise<CompilerOutput> {
+  const vyperPlatform = findVyperPlatform();
+  let vyperPath;
+  if (vyperPlatform) {
+    vyperPath = await getVyperExecutable(vyperRepoPath, vyperPlatform, version);
+  }
+
+  if (!vyperPath) {
+    throw new Error("Vyper path not found");
+  }
+
+  let compiled: string | undefined;
+  const inputStringified = JSON.stringify(vyperJsonInput);
+  try {
+    compiled = await asyncExecVyper(inputStringified, vyperPath);
+  } catch (error: any) {
+    if (error?.code === "ENOBUFS") {
+      throw new Error("Compilation output size too large");
+    }
+    logger.warn(error.message);
+    throw error;
+  }
+
+  if (!compiled) {
+    throw new Error("Compilation failed. No output from the compiler.");
+  }
+  const compiledJSON = JSON.parse(compiled);
+  const errorMessages = compiledJSON?.errors?.filter(
+    (e: any) => e.severity === "error",
+  );
+  if (errorMessages && errorMessages.length > 0) {
+    const error = new Error(
+      "Compiler error:\n " + JSON.stringify(errorMessages),
+    );
+    logger.error(error.message);
+    throw error;
+  }
+  return compiledJSON;
+}
+
+export async function getVyperExecutable(
+  vyperRepoPath: string,
+  platform: string,
+  version: string,
+): Promise<string | null> {
+  const fileName = `vyper.${version}.${platform}`;
+  const vyperPath = path.join(vyperRepoPath, fileName);
+  if (fs.existsSync(vyperPath) && validateVyperPath(vyperPath)) {
+    logger.debug("Found vyper binary", {
+      version,
+      vyperPath,
+      platform,
+    });
+    return vyperPath;
+  }
+
+  logger.debug("Downloading vyper", {
+    version,
+    vyperPath,
+    platform,
+  });
+  const success = await fetchAndSaveVyper(
+    platform,
+    vyperPath,
+    version,
+    fileName,
+  );
+  logger.debug("Downloaded vyper", {
+    version,
+    vyperPath,
+    platform,
+  });
+  if (success && !validateVyperPath(vyperPath)) {
+    logger.error("Cannot validate vyper", {
+      version,
+      vyperPath,
+      platform,
+    });
+    return null;
+  }
+  return success ? vyperPath : null;
+}
+
+function validateVyperPath(vyperPath: string): boolean {
```

Comment:
This is part of

> Create a common folder for the compilers code used by both lib-sourcify and server (idk if it fits in this new feature, imo yes because it's a lot of duplicated code now that we have also vyper stuff)